Mathlib/Order/RelIso/Basic.lean

```diff
@@ -0,0 +1,778 @@
+  `r a b → s (f a) (f b)`.
+* `RelEmbedding`: Relation embedding. A `RelEmbedding r s` is an embedding `f : α ↪ β` such that
+  `r a b ↔ s (f a) (f b)`.
+* `RelIso`: Relation isomorphism. A `RelIso r s` is an equivalence `f : α ≃ β` such that
+  `r a b ↔ s (f a) (f b)`.
+* `sum_lex_congr`, `prod_lex_congr`: Creates a relation homomorphism between two `sum_lex` or two
+  `prod_lex` from relation homomorphisms between their arguments.
+
+## Notation
+
+* `→r`: `rel_hom`
+* `↪r`: `rel_embedding`
+* `≃r`: `rel_iso`
+-/
+
+
+open Function
+
+universe u v w
+
+variable {α β γ δ : Type _} {r : α → α → Prop} {s : β → β → Prop} {t : γ → γ → Prop} {u : δ → δ → Prop}
+
+/-- A relation homomorphism with respect to a given pair of relations `r` and `s`
+is a function `f : α → β` such that `r a b → s (f a) (f b)`. -/
+structure RelHom {α β : Type _} (r : α → α → Prop) (s : β → β → Prop) where
+  toFun : α → β
+  map_rel' : ∀ {a b}, r a b → s (toFun a) (toFun b)
+#align rel_hom RelHom
+
+-- mathport name: «expr →r »
+infixl:25 " →r " => RelHom
+
+section
+
+/-- `rel_hom_class F r s` asserts that `F` is a type of functions such that all `f : F`
+satisfy `r a b → s (f a) (f b)`.
+
+The relations `r` and `s` are `out_param`s since figuring them out from a goal is a higher-order
+matching problem that Lean usually can't do unaided.
+-/
+class RelHomClass (F : Type _) {α β : outParam <| Type _} (r : outParam <| α → α → Prop)
+  (s : outParam <| β → β → Prop) extends FunLike F α fun _ => β where
+  map_rel : ∀ (f : F) {a b}, r a b → s (f a) (f b)
+#align rel_hom_class RelHomClass
+
+export RelHomClass (map_rel)
+
+end
+
+namespace RelHomClass
+
+variable {F : Type _}
+
+protected theorem is_irrefl [RelHomClass F r s] (f : F) : ∀ [IsIrrefl β s], IsIrrefl α r
+  | ⟨H⟩ => ⟨fun a h => H _ (map_rel f h)⟩
+#align rel_hom_class.is_irrefl RelHomClass.is_irrefl
+
+protected theorem is_asymm [RelHomClass F r s] (f : F) : ∀ [IsAsymm β s], IsAsymm α r
+  | ⟨H⟩ => ⟨fun a b h₁ h₂ => H _ _ (map_rel f h₁) (map_rel f h₂)⟩
+#align rel_hom_class.is_asymm RelHomClass.is_asymm
+
+protected theorem acc [RelHomClass F r s] (f : F) (a : α) : Acc s (f a) → Acc r a := by
+  generalize h : f a = b
+  intro ac
+  induction' ac with _ H IH generalizing a
+  subst h
+  exact ⟨_, fun a' h => IH (f a') (map_rel f h) _ rfl⟩
+#align rel_hom_class.acc RelHomClass.acc
+
+protected theorem well_founded [RelHomClass F r s] (f : F) : ∀ h : WellFounded s, WellFounded r
+  | ⟨H⟩ => ⟨fun a => RelHomClass.acc f _ (H _)⟩
+#align rel_hom_class.well_founded RelHomClass.well_founded
+
+end RelHomClass
+
+namespace RelHom
+
+instance : RelHomClass (r →r s) r s where
+  coe o := o.toFun
+  coe_injective' f g h := by
+    cases f
+    cases g
+    congr
+  map_rel := map_rel'
+
+/-- Auxiliary instance if `rel_hom_class.to_fun_like.to_has_coe_to_fun` isn't found -/
+instance : CoeFun (r →r s) fun _ => α → β :=
+  ⟨fun o => o.toFun⟩
+
```

Comment:
Is this really needed? It seems likely that this is causing the delaboration issue. If we can't remove it, perhaps try adding `instance [coe] RelHom.toFun`.